### PR TITLE
Check RDom::where predicates for race conditions

### LINKED ISSUE #6808: The check for race conditions doesn't consider where clauses
Halide should require allow_race_conditions() to be willing to compile the following code, but it doesn't. The where clause introduces a race.

```
    Func f;
    Var x;
    RDom r(0, 10);
    f(x) = 1;
    r.where(f(0) == 1);
    f(r) = 2;

    f.update().parallel(r);
```

ATTN @alexreinking

### PR DESCRIPTION
**Pull request by alexreinking**
Fixes #6808

### CODE REVIEW

File: src/Func.cpp [hunk 1]
```diff
@@ -386,6 +386,22 @@ void Stage::set_dim_type(const VarOrRVar &var, ForType t) {
...
             dim.for_type = t;
 
+            // Check that RDom::where is not data-dependent on the current func.
+            // This check is fairly conservative.
+            if (var.is_rvar && is_parallel(t) && !definition.schedule().allow_race_conditions()) {
+                const string &func_name = function.name();
+                const auto &rdom = var.rvar.domain();
+                if (rdom.defined()) {
+                    user_assert(!expr_uses_var(rdom.predicate(), func_name))
+                        << "In scheudle for " << name() << ", marking var "
```
Comment (abadams) [inline src/Func.cpp:396]: scheudle

**Comment by abadams** [comment]
It might be simpler and more precise to visit the rdom predicate when sniffing for loads here: https://github.com/halide/Halide/blob/main/src/ParallelRVar.cpp#L93

File: src/Func.cpp (hunk 1 above)
Comment (alexreinking) [reply src/Func.cpp]: Fixed

**Review by abadams — APPROVED** [review]

**Comment by alexreinking** [comment]
Windows GPU atomics failure is unrelated.

**Pull request merged**